content/articles/do-we-need-a-database-when-we-have-apache-kafka/index.md

```diff
@@ -0,0 +1,66 @@
+
+The next step is optimization. In Optimization, the database system modifies the user's query to generate an efficient evaluation plan that minimizes cost. After optimization, evaluation begins and it entails using relational algebra expressions to evaluate an operation. Finally, the query execution engine generates the output from the given query and subjects it to processing. Processing in a database includes transforming data into a useful output. The output takes the form of reports and documents. In a database, the output obtained from processing is simpler.
+
+### Transactions in Kafka
+
+The handling of Apache Kafka transactions occurs on the producer's side instead of the consumer's side. The consumer reads while the producer takes up the role of coordinating the transaction. This approach reduces the performance overload of the consumer but at the cost of the producer. On a distributed database system, the message broker/producer manages the transactions by default. Such transactions are known as [broker-coordinated](https://www.ibm.com/docs/en/integration-bus/9.0.0?topic=behavior-message-flow-transactions) transactions.
+
+Kafka and database transactions share similarities because they possess the ACID capability. They guarantee atomicity, consistency, isolation, and durability. Atomicity refers to all the elements that constitute a complete database transaction. Consistency refers to the rules that maintain data points in a correct state following a transaction. Isolation separates the effect of a transaction from others until it commits, and durability ensures that data changes become permanent after they are committed.
+
+### Kafka and ACID guarantees
+
+Kafka, like databases, guarantees ACID. Its transactional systems implement atomicity, consistency, isolation, and durability. A transaction behaves like a single unit that can either succeed or completely fail. Observers can see all the effects of a transaction at the same moment. Also, once the system commits a given transaction, it remains in that state even in cases of downtimes.
+
+Even so, it is notable that Kafka faces the potential challenge of a failure that can break ACID compliance. For instance, in the case of an under-replicated partition accompanied by an unforeseen power shortage, data loss would be inevitable. Such a transaction will lose its integrity.
+
+### Conclusion
+
+Databases imply an organized collection of data that is electronically accessible on a computer system. Data collection remains significant to businesses. Most databases rely on conventional approaches to grant data access, which does not befit current big data firms like social media platforms.
+
+Kafka has been the solution to the challenges that came with using databases. It is a streaming service platform that offers database-like functions. Modern firms are now favoring this system because it is scalable and fault-tolerant. It is more of a hybrid database system.
+
```

Comment:
Here, you can summarize the key takeaways discussed in the article (in point form) to help you answer the topic question of the article (Do we need a Database when we have Apache Kafka?). I think this is important because it is the main focus of the article. This will help readers satisfy the curiosity gained at the beginning of the article.